File: services/translation-worker/vantara_worker/ocr.py
"""قراءة النص الأصلي داخل كل منطقة.

  * اللاتيني (إنجليزي وما يشبهه): PP-OCRv5 mobile rec (Apache-2.0) على كل سطر.
    الأسطر تُشتق من قناع الحروف نفسه (إسقاط أفقي داخل صندوق النص) لا من
    نموذج آخر: ما اعتبره القناع حرفًا هو ما يُقرأ، فلا سطر يضيع.
  * الياباني: manga-ocr (Apache-2.0، ONNX int8) على المنطقة كاملة؛ يفهم العمودي.

القياس على الصفحات الحقيقية: 10 من 10 فقاعات إنجليزية قُرئت حرفيًا بثقة ≥ 0.93،
في 0.05–0.10 ثانية للصفحة.

القراءة هنا «مسودّة»: Luna ترى الصورة أيضًا وتصحّح ما أخطأه OCR، لكن وجود
نص مقروء محليًّا يعطي عتبة أمان: منطقة بلا حروف مقروءة لا تُبيَّض.
"""

from __future__ import annotations

from dataclasses import dataclass

import cv2
import numpy as np

from . import models
from .vision.common import Box, make_session


@dataclass
class OcrLine:
    box: Box
    text: str
    confidence: float


@dataclass
class OcrResult:
    text: str
    confidence: float
    lines: list[OcrLine]


def split_lines(glyph: np.ndarray, box: Box, min_gap: int = 3) -> list[Box]:
    """أسطر النص من الإسقاط الأفقي لقناع الحروف داخل الصندوق."""
    x1, y1, x2, y2 = box
    sub = glyph[y1:y2, x1:x2] > 0
    if not sub.any():
        return []
    rows = sub.sum(axis=1)
    lines: list[Box] = []
    inside = False
    start = 0
    gap = 0
    for y, v in enumerate(rows):
        if v > 0:
            if not inside:
                inside, start = True, y
            gap = 0
        elif inside:
            gap += 1
            if gap >= min_gap:
                lines.append((y1 + start, y1 + y - gap))
                inside = False
    if inside:
        lines.append((y1 + start, y1 + len(rows)))
    out: list[Box] = []
    heights = [b - a for a, b in lines]
    typical = float(np.median(heights)) if heights else 0
    for a, b in lines:
        if b - a < max(4, typical * 0.35):  # نقطة أو شرطة وحدها ليست سطرًا
            continue
        cols = np.where(sub[a - y1 : b - y1].any(axis=0))[0]
        out.append((x1 + int(cols[0]), a, x1 + int(cols[-1]) + 1, b))
    return out


class LatinOcr:
    """PP-OCRv5 CTC recognition، سطرٌ سطرًا."""

    HEIGHT = 48

    def __init__(self) -> None:
        self.session = make_session(str(models.ensure("ppocr_en_rec")), threads=2)
        self.input = self.session.get_inputs()[0].name
        with open(models.ensure("ppocr_en_dict"), encoding="utf-8") as f:
            self.chars = [line.rstrip("\n") for line in f]

    def recognize(self, bgr_line: np.ndarray) -> tuple[str, float]:
        h, w = bgr_line.shape[:2]
        if h < 4 or w < 4:
            return "", 0.0
        W = max(16, int(np.ceil(self.HEIGHT * w / h)))
        x = cv2.resize(bgr_line, (W, self.HEIGHT), interpolation=cv2.INTER_LINEAR).astype(np.float32) / 255.0
        x = ((x - 0.5) / 0.5).transpose(2, 0, 1)[None]
        logits = self.session.run(None, {self.input: x})[0][0]
        vocab = ["", *self.chars, " "] if logits.shape[-1] == len(self.chars) + 2 else ["", *self.chars]
        if logits.max() > 1.0 or logits.min() < 0.0:
            e = np.exp(logits - logits.max(-1, keepdims=True))
            logits = e / e.sum(-1, keepdims=True)
        idx = logits.argmax(-1)
        chars: list[str] = []
        confs: list[float] = []
        last = -1
        for t, i in enumerate(idx):
            if i != 0 and i != last and i < len(vocab):
                chars.append(vocab[i])
                confs.append(float(logits[t, i]))
            last = int(i)
        return "".join(chars).strip(), float(np.mean(confs)) if confs else 0.0

    def read(self, rgb: np.ndarray, glyph: np.ndarray, box: Box, pad: int = 4) -> OcrResult:
        H, W = rgb.shape[:2]
        bgr = cv2.cvtColor(rgb, cv2.COLOR_RGB2BGR)
        lines: list[OcrLine] = []
        for lx1, ly1, lx2, ly2 in split_lines(glyph, box):
            crop = bgr[max(0, ly1 - pad) : min(H, ly2 + pad), max(0, lx1 - pad) : min(W, lx2 + pad)]
            text, conf = self.recognize(crop)
            if text:
                lines.append(OcrLine((lx1, ly1, lx2, ly2), text, conf))
        text = " ".join(line.text for line in lines)
        conf = float(np.mean([line.confidence for line in lines])) if lines else 0.0
        return OcrResult(text, conf, lines)


class MangaOcr:
    """manga-ocr: ترميز الصورة 224×224 ثم فكّ تسلسلي greedy."""

    def __init__(self) -> None:
        self.encoder = make_session(str(models.ensure("mangaocr_encoder")), threads=2)
        self.decoder = make_session(str(models.ensure("mangaocr_decoder")), threads=2)
        with open(models.ensure("mangaocr_vocab"), encoding="utf-8") as f:
            self.vocab = f.read().splitlines()
        self.enc_in = self.encoder.get_inputs()[0].name
        names = [i.name for i in self.decoder.get_inputs()]
        self.dec_tokens = next((n for n in names if "input_ids" in n or "token" in n), names[0])
        self.dec_hidden = next((n for n in names if "encoder" in n), names[-1])

    def read(self, rgb: np.ndarray, glyph: np.ndarray, box: Box, pad: int = 6) -> OcrResult:
        H, W = rgb.shape[:2]
        x1, y1, x2, y2 = box
        crop = rgb[max(0, y1 - pad) : min(H, y2 + pad), max(0, x1 - pad) : min(W, x2 + pad)]
        gray = cv2.cvtColor(cv2.cvtColor(crop, cv2.COLOR_RGB2GRAY), cv2.COLOR_GRAY2RGB)
        im = cv2.resize(gray, (224, 224), interpolation=cv2.INTER_LINEAR).astype(np.float32) / 255.0
        x = ((im - 0.5) / 0.5).transpose(2, 0, 1)[None]
        hidden = self.encoder.run(None, {self.enc_in: x})[0]
        tokens = [2]
        for _ in range(300):
            logits = self.decoder.run(None, {self.dec_tokens: np.array([tokens], dtype=np.int64), self.dec_hidden: hidden})[0]
            nxt = int(np.argmax(logits[0, -1]))
            tokens.append(nxt)
            if nxt == 3:
                break
        text = "".join(self.vocab[t] for t in tokens if t >= 5 and t < len(self.vocab))
        text = "".join(text.split()).replace("…", "...")
        return OcrResult(text, 1.0 if text else 0.0, [OcrLine(box, text, 1.0 if text else 0.0)])

File: android/app/src/main/kotlin/com/vantara/plugins/translation/Ocr.kt
package com.vantara.plugins.translation

import ai.onnxruntime.OrtSession
import java.io.File
import kotlin.math.ceil

data class OcrLine(val box: Box, val text: String, val confidence: Float)
data class OcrResult(val text: String, val confidence: Float, val lines: List<OcrLine>)

/**
 * قراءة النص اللاتيني سطرًا سطرًا: PP-OCRv5 mobile rec (CTC).
 * الأسطر تُشتق من الإسقاط الأفقي لقناع الحروف داخل صندوق النص.
 */
class LatinOcr(model: File, dict: File) {
    private val session: OrtSession = Ort.open(model, threads = 2)
    private val chars: List<String> = dict.readLines(Charsets.UTF_8)
    private val inputName = session.inputNames.first()
    private val height = 48

    /** أسطر النص من الإسقاط الأفقي لقناع الحروف داخل الصندوق. */
    fun splitLines(glyph: ByteMask, box: Box, minGap: Int = 3): List<Box> {
        val rows = IntArray(box.h)
        for (y in 0 until box.h) {
            var n = 0
            val yy = box.y1 + y
            if (yy < 0 || yy >= glyph.height) continue
            for (x in maxOf(0, box.x1) until minOf(glyph.width, box.x2)) if (glyph[x, yy].toInt() != 0) n++
            rows[y] = n
        }
        val spans = ArrayList<IntArray>()
        var inside = false; var start = 0; var gap = 0
        for (y in rows.indices) {
            if (rows[y] > 0) { if (!inside) { inside = true; start = y }; gap = 0 }
            else if (inside) { gap++; if (gap >= minGap) { spans.add(intArrayOf(start, y - gap + 1)); inside = false } }
        }
        if (inside) spans.add(intArrayOf(start, rows.size))
        val heights = spans.map { it[1] - it[0] }.sorted()
        val typical = if (heights.isEmpty()) 0 else heights[heights.size / 2]
        val out = ArrayList<Box>()
        for ((a, b) in spans.map { it[0] to it[1] }) {
            if (b - a < maxOf(4, (typical * 0.35).toInt())) continue
            var x0 = Int.MAX_VALUE; var x1 = -1
            for (y in a until b) for (x in maxOf(0, box.x1) until minOf(glyph.width, box.x2)) if (glyph[x, box.y1 + y].toInt() != 0) { if (x < x0) x0 = x; if (x > x1) x1 = x }
            if (x1 < 0) continue
            out.add(Box(x0, box.y1 + a, x1 + 1, box.y1 + b))
        }
        return out
    }

    /** يقرأ سطرًا واحدًا (قطعة RGB). يرجع النص ومتوسط ثقة الحروف. */
    fun recognize(line: RgbImage): Pair<String, Float> {
        if (line.width < 4 || line.height < 4) return "" to 0f
        val w = maxOf(16, ceil(height * line.width.toFloat() / line.height).toInt())
        val small = line.resize(w, height)
        // BGR وتطبيع إلى [-1,1]: النموذج مدرَّب على ترتيب OpenCV
        val plane = w * height
        val x = FloatArray(3 * plane)
        for (i in 0 until plane) {
            val r = (small.data[i * 3].toInt() and 0xff) / 255f
            val g = (small.data[i * 3 + 1].toInt() and 0xff) / 255f
            val b = (small.data[i * 3 + 2].toInt() and 0xff) / 255f
            x[i] = (b - 0.5f) / 0.5f; x[plane + i] = (g - 0.5f) / 0.5f; x[2 * plane + i] = (r - 0.5f) / 0.5f
        }
        val input = Ort.tensor(x, 1, 3, height.toLong(), w.toLong())
        val sb = StringBuilder()
        val confs = ArrayList<Float>()
        session.run(mapOf(inputName to input)).use { res ->
            @Suppress("UNCHECKED_CAST")
            val logits = (res[0].value as Array<Array<FloatArray>>)[0] // T × C
            val c = logits[0].size
            val vocab = if (c == chars.size + 2) listOf("") + chars + listOf(" ") else listOf("") + chars
            var last = -1
            for (t in logits.indices) {
                val row = logits[t]
                var best = 0; var bestV = row[0]
                for (i in 1 until row.size) if (row[i] > bestV) { bestV = row[i]; best = i }
                if (best != 0 && best != last && best < vocab.size) {
                    // softmax للثقة إن كانت لوغاريتمية
                    var maxV = row[0]; for (v in row) if (v > maxV) maxV = v
                    var sum = 0.0; for (v in row) sum += Math.exp((v - maxV).toDouble())
                    val p = if (row.any { it > 1f || it < 0f }) (Math.exp((bestV - maxV).toDouble()) / sum).toFloat() else bestV
                    sb.append(vocab[best]); confs.add(p)
                }
                last = best
            }
        }
        input.close()
        return sb.toString().trim() to (if (confs.isEmpty()) 0f else confs.average().toFloat())
    }

    fun read(img: RgbImage, glyph: ByteMask, box: Box, pad: Int = 4): OcrResult {
        val lines = ArrayList<OcrLine>()
        for (l in splitLines(glyph, box)) {
            val crop = img.crop(maxOf(0, l.x1 - pad), maxOf(0, l.y1 - pad), minOf(img.width, l.x2 + pad), minOf(img.height, l.y2 + pad))
            val (text, conf) = recognize(crop)
            if (text.isNotEmpty()) lines.add(OcrLine(l, text, conf))
        }
        val text = lines.joinToString(" ") { it.text }
        val conf = if (lines.isEmpty()) 0f else lines.map { it.confidence }.average().toFloat()
        return OcrResult(text, conf, lines)
    }

    fun close() = session.close()
}
